RFCs/FS-1063-support-letbang-andbang-for-applicative-functors.md

```diff
@@ -0,0 +1,750 @@
+
+With applicative computation expressions, we can write more computations than before (there are more contexts which meet the requirements for applicative computation expressions than the existing monadic ones), and we can write more efficient computations (the requirements of applicatives rule out needing to support some potentially expensive operations).
+
+For example, [Pauan points out](https://github.com/fsharp/fslang-suggestions/issues/579#issuecomment-310799948) that we can write a computation expression for `Observable`s that [avoids unnecessary resubscriptions](https://github.com/fsharp/fslang-suggestions/issues/579#issuecomment-310854419):
+
+```fsharp
+// Outputs a + b, which is recomputed every time foo or bar outputs a new value,
+// avoiding any unnecessary resubscriptions
+observable {
+  let! a = foo
+  and! b = bar
+  return a + b
+}
+```
+
+## Why Applicatives?
+
+[Applicative functors](https://en.wikipedia.org/wiki/Applicative_functor) sit, on the spectrum of flexibility vs. predictability, somewhere between [functors](https://en.wikipedia.org/wiki/Functor#Computer_implementations) (i.e. types which support `Map`), and [monads](https://en.wikipedia.org/wiki/Monad_(functional_programming)) (i.e. types which support `Bind`, which currently underpin computation expressions).
+
+If we consider `Bind : M<'T> * ('T -> M<'U>) -> M<'U>`, we can see that the second element of the input is a function that requires a value to create the resulting "wrapped value". This means the argument to `Bind` has the power to completely change the context of the result based on the value seen (e.g. to create and destroy `Observable` subscriptions), but it also means that the expression builder can predict much less about what the given function will decide to do, and hence has fewer outcomes that it can rule out and potentially optimise away.
+
+In contrast, `Apply : M<'T -> 'U> * M<'T> -> M<'U>` only needs a wrapped function, which is something we have whilst building our computation and not something that can be controlled by the values at come later. This removes some flexibility to drastically alter the shape of the context in response to values seen later, but means that the computation expression builder now knows much more about what can or cannot happen after construction, and hence can make intelligent decisions off the back of that (e.g. to avoid unsubscribing only to immediately resubscribe, or to perhaps run two operations in parallel because it knows there can be no dependencies between them).
```

Comment:
`Apply` feels to me to specific for the nature of computation expressions. I lean more to the `Merge` + `Map` proposed in [The F# Computation Expression Zoo](http://tomasp.net/academic/papers/computation-zoo/). I think it is at least worth adding a section discussing this option and showing expansions and use cases as to why `Apply` is the preferred option here.

Reply:
NOTE: I wanted to note this as I suspect `Map` and `Merge` are slightly more general and thus more in the spirit of F# computation expressions.

Reply:
It was my understanding that `Map` and `Merge` offer precisely the power of `Apply` and `Return`. However, `Merge` introduces the notion of tuples, which I find inelegant, but more importantly, `Apply` is the standard formulation that one would see in the wider literature.

Reply:
I think the big challenge with computation expressions is remembering that `Bind` and `Return` are not exactly equal to the `bind` and `return` of monads. It's common to fall into the trap of thinking of them in this way. I don't think it a _wrong_ interpretation; it's just that CEs are _not_ monad builders or applicative builders, though you can encode such constructs within them.

I don't think `Apply` is a bad direction; however, I think it worth expressing the expansions side-by-side and thinking about the tradeoffs. For example, if `Apply` and `Return` offer a wider range of expressions within CEs, then we should use that. If `Map` and `Merge` offer a wider expression, then we should use that. Category should only enter the picture to ensure that such constructs _can_ be encoded. They should not, imo, be the sole purpose for expanding the builder syntax. I think this is in keeping with the spirit of how they have been implemented to date.

Reply:
Applicatives culturally did not directly come from category theory, they came out of programming and are merely associated with category theory through functors. That's why I put "proper" in quotes, because *technically* there is no such thing as an applicative in category theory. It's much more useful to us programmers and less interesting to pure mathematicians so it gets a special name for us. I think breaking it out into map and merge add indirection in a way that makes it less accessible, additionally it's more work. Map is already being considered in a different RFC but considering merge. Merge lets us take a tuple where each term is the same container type, and return a container type that contains a tuple. Apply lets me sidestep the tuple dance and stay in whatever container type I'm in. I can see the value of using a monoidal conceptual model in Tomas' terms "affect what (applicative) computations are evaluated as part of the computation.". However I think defining an intuitive syntax and good UX be a much more complex and involved process, that's why I think merge and map computation expressions should be its own RFC. It looks like Tomas has already put some effort into designing such a thing, but has not given any good applications outside of formlets. Your standard applicative functor is much simpler syntactically, I would hate for merge/map to become something that prevents us from ever actually getting Apply, something which has many more uses and is easier to implement and test.


http://tryjoinads.org/index.html?computations/applicative.html
Apply 
``` fsharp
val apply : F<'T1 -> 'T2> -> F<'T1> -> F<'T2>

val merge:  F<'T1> * F<'T2> -> F<'T1 * 'T2>
```

Reply:
That is a good explanation. Thanks!